Parallelize module loading in POSIX dyld code

zhyty opened the pull request:
This patch improves LLDB launch time on Linux machines for **preload scenarios**, particularly for executables with a lot of shared library dependencies (or modules). Specifically:
* Launching a binary with `target.preload-symbols = true` 
* Attaching to a process with `target.preload-symbols = true`.
It's completely controlled by a new flag added in the first commit `plugin.dynamic-loader.posix-dyld.parallel-module-load`, which *defaults to false*. This was inspired by similar work on Darwin #110646. 

Some rough numbers to showcase perf improvement, run on a very beefy machine:
* Executable with ~5600 modules: baseline 45s, improvement 15s
* Executable with ~3800 modules: baseline 25s,  improvement 10s
* Executable with ~6650 modules: baseline 67s, improvement 20s
* Executable with ~12500 modules: baseline 185s, improvement 85s
* Executable with ~14700 modules: baseline 235s, improvement 120s
A lot of targets we deal with have a *ton* of modules, and unfortunately we're unable to convince other folks to reduce the number of modules, so performance improvements like this can be very impactful for user experience.

This patch achieves the performance improvement by parallelizing `DynamicLoaderPOSIXDYLD::RefreshModules` for the launch scenario, and `DynamicLoaderPOSIXDYLD::LoadAllCurrentModules` for the attach scenario. The commits have some context on their specific changes as well -- hopefully this helps the review.

# More context on implementation

We discovered the bottlenecks by via `perf record -g -p <lldb's pid>` on a Linux machine. With an executable known to have 1000s of shared library dependencies, I ran
```
(lldb) b main
(lldb) r
# taking a while
```
and showed the resulting perf trace (snippet shown)
```
Samples: 85K of event 'cycles:P', Event count (approx.): 54615855812
  Children      Self  Command          Shared Object              Symbol
-   93.54%     0.00%  intern-state     libc.so.6                  [.] clone3
     clone3
     start_thread
     lldb_private::HostNativeThreadBase::ThreadCreateTrampoline(void*)                                                                           r
     std::_Function_handler<void* (), lldb_private::Process::StartPrivateStateThread(bool)::$_0>::_M_invoke(std::_Any_data const&)
     lldb_private::Process::RunPrivateStateThread(bool)                                                                                          n
   - lldb_private::Process::HandlePrivateEvent(std::shared_ptr<lldb_private::Event>&)
      - 93.54% lldb_private::Process::ShouldBroadcastEvent(lldb_private::Event*)
         - 93.54% lldb_private::ThreadList::ShouldStop(lldb_private::Event*)
            - lldb_private::Thread::ShouldStop(lldb_private::Event*)                                                                             *
               - 93.53% lldb_private::StopInfoBreakpoint::ShouldStopSynchronous(lldb_private::Event*)                                            t
                  - 93.52% lldb_private::BreakpointSite::ShouldStop(lldb_private::StoppointCallbackContext*)                                     i
                       lldb_private::BreakpointLocationCollection::ShouldStop(lldb_private::StoppointCallbackContext*)                           k
                       lldb_private::BreakpointLocation::ShouldStop(lldb_private::StoppointCallbackContext*)                                     b
                       lldb_private::BreakpointOptions::InvokeCallback(lldb_private::StoppointCallbackContext*, unsigned long, unsigned long)    i
                       DynamicLoaderPOSIXDYLD::RendezvousBreakpointHit(void*, lldb_private::StoppointCallbackContext*, unsigned long, unsigned lo
                     - DynamicLoaderPOSIXDYLD::RefreshModules()                                                                                  O
                        - 93.42% DynamicLoaderPOSIXDYLD::RefreshModules()::$_0::operator()(DYLDRendezvous::SOEntry const&) const                 u
                           - 93.40% DynamicLoaderPOSIXDYLD::LoadModuleAtAddress(lldb_private::FileSpec const&, unsigned long, unsigned long, bools
                              - lldb_private::DynamicLoader::LoadModuleAtAddress(lldb_private::FileSpec const&, unsigned long, unsigned long, boos
                                 - 83.90% lldb_private::DynamicLoader::FindModuleViaTarget(lldb_private::FileSpec const&)                        o
                                    - 83.01% lldb_private::Target::GetOrCreateModule(lldb_private::ModuleSpec const&, bool, lldb_private::Status*
                                       - 77.89% lldb_private::Module::PreloadSymbols()
                                          - 44.06% lldb_private::Symtab::PreloadSymbols()
                                             - 43.66% lldb_private::Symtab::InitNameIndexes()
...
```
We saw that majority of time was spent in `RefreshModules`, with the main culprit within it `LoadModuleAtAddress` which eventually calls `PreloadSymbols`. 

At first, `DynamicLoaderPOSIXDYLD::LoadModuleAtAddress` appears fairly independent -- most of it deals with different files and then getting or creating Modules from these files. The portions that aren't independent seem to deal with ModuleLists, which appear concurrency safe. There were members of `DynamicLoaderPOSIXDYLD` I had to synchronize though: namely `m_loaded_modules` which `DynamicLoaderPOSIXDYLD` maintains to map its loaded modules to their link addresses. Without synchronizing this, I ran into SEGFAULTS and other issues when running `check-lldb`. I also locked the assignment and comparison of `m_interpreter_module`, which may be unnecessary.

# Alternate implementations

When creating this patch, another implementation I considered was directly background-ing the call to `Module::PreloadSymbol` in `Target::GetOrCreateModule`. It would have the added benefit of working across platforms generically, and appeared to be concurrency safe. It was done via `Debugger::GetThreadPool().async` directly. However, there were a ton of concurrency issues, so I abandoned that approach for now.

# Testing

With the feature active, I tested via `ninja check-lldb` on both Debug and Release builds several times (~5 or 6 altogether?), and didn't spot additional failing or flaky tests.

I also tested manually on several different binaries, some with around 14000 modules, but just basic operations: launching, reaching main, setting breakpoint, stepping, showing some backtraces.

I've also tested with the flag off just to make sure things behave properly synchronously.

Commit:
Add option for enabling/disabling parallel load

Add a setting to gate the new parallel dynamic library loading feature in the next diff. I followed the example of `lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp` and `lldb/source/Plugins/StructuredData/DarwinLog/StructuredDataDarwinLog.cpp` to create a scoped setting for the new feature.

NOTE: this setting defaults to FALSE.

Users should be able to enable the new feature as easily as adding a line to their .lldbinit file, or manually setting it in their session.

```
(lldb) apropos "POSIX dynamic"
No commands found pertaining to 'POSIX dynamic'. Try 'help' to see a complete list of debugger commands.

The following settings variables may relate to 'POSIX dynamic':
  plugin.dynamic-loader.posix-dyld.parallel-module-load -- Enable experimental loading of modules in parallel for the POSIX dynamic loader.
(lldb) settings show plugin.dynamic-loader.posix-dyld.parallel-module-load
plugin.dynamic-loader.posix-dyld.parallel-module-load (boolean) = false
(lldb) settings set plugin.dynamic-loader.posix-dyld.parallel-module-load true
(lldb) settings show plugin.dynamic-loader.posix-dyld.parallel-module-load
plugin.dynamic-loader.posix-dyld.parallel-module-load (boolean) = true
```

Commit:
parallelize POSIX dyld RefreshModules

This diff parallelizes `DynamicLoaderPOSIXDYLD::RefreshModules`, which speeds up module loading on Linux. The major benefit of this is we can speed up symbol table indexing and parsing, which is the biggest bottleneck for targets which dynamically link many shared libraries.

This speedup is only noticeable when **preloading** symbols. This is when `target.preload-symbols` is `true`, which is the default Meta. The symbol preload option tells the debugger to fully load all of the symbol tables when modules are loaded, as opposed to lazily loading when symbols are requested.

Initially, I discovered the specific bottleneck by using the Linux `perf` tool. I saw that ~93% of samples were in `RefreshModules`, and mainly in `LoadModuleAtAddress` and `PreloadSymbols`.

`LoadModuleAtAddress` appears independent and parallelize-able at first. The main issue is `DynamicLoaderPOSIXDYLD` maintains a map of loaded modules to their link addresses via `m_loaded_modules`. Modifying and reading to this map isn't thread-safe, so this diff also includes accessor methods that protect the map in the multithreaded context. Luckily, the critical section of modifying or reading from the map isn't super costly, so the contention doesn't appear to negatively impact performance.

I tested with some larger projects with up to 15000 modules, and
found significant performance improvements. Typically, I was seeing 2-3X
launch speed increases, where "launch speed" is starting the binary and
reaching `main`.

I manually ran `ninja check-lldb` several times, and compared with the baseline. At this point, we're not seeing any new failures or new unresolved tests.

Commit:
parallelize dyld for attach scenario

JDevlieghere reviewed (commented):
This looks like the equivalent of #110646 but for Linux and the overall approach looks good to me. 

My only real concern is about the setting and it being **off** by default. I would prefer that this is **on** by default and the setting is there as an escape hatch in case there's an unexpected issue. I'd like to understand what's holding that back?

Building on that, if the setting is on by default, then instead of creating a new setting per plugin, I'd like to hoist it into target and have one setting that controls this behavior for all of them. I expect it to be rare that you'd want to turn this off for one plugin but not the other. Unless we are a 100% sure we're going to take away the setting (in which case the setting should be experimental[1]), it's easier to keep it around, for example to make it easier to debug or quickly toggle it and see if a concurrency issue disappears. 

[1] See `plugin.dynamic-loader.darwin.experimental.enable-parallel-image-load`. The experimental prefix allows you to set that option without it existing for when it either gets promoted to a real setting (without the experimental) or alternative it goes away.

JDevlieghere commented on lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLDProperties.td:
```diff
@@ -0,0 +1,8 @@
+include "../../../../include/lldb/Core/PropertiesBase.td"
+
+let Definition = "dynamicloaderposixdyld" in {
+  def ParallelModuleLoad: Property<"parallel-module-load", "Boolean">,
```
If we settle on keeping the per-plugin setting, this should be consistent with `plugin.dynamic-loader.darwin.experimental.enable-parallel-image-load`, i.e. be under experimental and use `enable-parallel-image-load` (even though I like your setting without the enable a little more). If we hoist it to target, then I'd like to go with `parallel-module-load`.

zhyty replied:
I don't feel strongly about keeping it as a plugin-specific setting or hoisting it to target and setting it generically. I missed the naming for the Darwin-equivalent though -- will rename depending on what we choose. 

I initially opted for the plugin-scoped setting since I wanted to make the setting as specific as possible, because I wasn't sure how many other platforms had a specific parallel loading flag like Darwin.

zhyty commented on lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp:
```diff
@@ -423,34 +493,66 @@ void DynamicLoaderPOSIXDYLD::RefreshModules() {
+          new_modules.Append(module_sp);
+        };
+
+    // Loading modules in parallel tends to be faster, but is still unstable.
```
There are no known issues. That comment is a leftover from the development process + me being too conservative. I felt at the time until we had tons of users testing it, I would call it "unstable". But, that's confusing in hindsight, so I'll remove the comment.

zhyty commented:
> This looks like the equivalent of #110646 but for Linux and the overall approach looks good to me.
> 
> My only real concern is about the setting and it being **off** by default. I would prefer that this is **on** by default and the setting is there as an escape hatch in case there's an unexpected issue. I'd like to understand what's holding that back?
> 
> Building on that, if the setting is on by default, then instead of creating a new setting per plugin, I'd like to hoist it into target and have one setting that controls this behavior for all of them. I expect it to be rare that you'd want to turn this off for one plugin but not the other. Unless we are a 100% sure we're going to take away the setting (in which case the setting should be experimental[1]), it's easier to keep it around, for example to make it easier to debug or quickly toggle it and see if a concurrency issue disappears.
> 
> [1] See `plugin.dynamic-loader.darwin.experimental.enable-parallel-image-load`. The experimental prefix allows you to set that option without it existing for when it either gets promoted to a real setting (without the experimental) or alternative it goes away.

Thanks for the review.

Yes, this PR was inspired by #110646 (I didn't have the PR number on hand, will link in the description). 

Internally, we were planning to have it on by default as well, but we felt that maybe it would be easier to land in open source if we had it off by default. I'm in favor of having it on as well, having no known issues.

JDevlieghere commented:
Great, if that's the case let's turn it on by default, and go with the shared setting at the target level. We can migrate over the Darwin plugin in a follow-up PR.

Commit:
move plugin.dynamic-loader.posix-dyld.parallel-module-load to
target.parallel-module-load

Commit:
switch usage of std shared_mutex with rwmutex, which is more portable

zhyty commented on lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp:
```diff
@@ -423,34 +493,66 @@ void DynamicLoaderPOSIXDYLD::RefreshModules() {
+            }
+          }
+
+          loaded_modules.AppendIfNeeded(module_sp);
+          new_modules.Append(module_sp);
```
Yes, so the threads don't read from `new_modules` during loading, and `new_modules` is only used after all modules have been loaded in parallel in `ModulesDidLoad`, which does some more initialization synchronously.

I added a comment to explain this -- hopefully the comment is clear enough.

zhyty commented:
@JDevlieghere in these 2 commits, I hoisted the flag out to `target`, fixed/added comments where appropriate, and switched the `shared_mutex` to `RWMutex`.

JDevlieghere reviewed (approved):
LGTM!

zhyty committed:
Merge branch 'main' into posix-parallel-dyld

JDevlieghere commented:
Forgot to ask here earlier, but would you also be able to add a release note in `llvm-project/llvm/docs/ReleaseNotes.md`? Thanks!
